File: sparkcql-synthea/src/test/scala/gov/va/sparkcql/synthea/SyntheaSourceAdapterTest.scala
package gov.va.sparkcql.synthea

import gov.va.sparkcql.TestBase
import org.apache.spark.sql.{SparkSession, Dataset, DataFrame, Row}
import org.apache.spark.sql.functions._
import org.hl7.elm.r1.Code
import gov.va.sparkcql.core.session.SparkCqlSession
import gov.va.sparkcql.core.adapter.source.FileSource
import gov.va.sparkcql.fhir.{FhirDataType, FhirModel}
import gov.va.sparkcql.core.model.DataType

class SyntheaSourceAdapterTest extends TestBase {
  
  val sparkcql = {
    SparkCqlSession.build(spark)
      .withSource(FileSource("./src/test/resources/cql"))
      .withSource(SyntheaSource(PopulationSize.PopulationSize10))
      .withModel(FhirModel())
      .create()
  }

  "A SyntheaSourceAdapter" should "dynamically retrieve resources" in {
    assert(sparkcql.retrieve(FhirDataType("Encounter")).get.head().getAs[String]("status") != null)
    assert(sparkcql.retrieve(FhirDataType("Condition")).get.head().getAs[String]("id") != null)
  }

  "A SyntheaDataProvider" should "lazily initialize" in {
    // val provider = new SyntheaDataProvider(PopulationSize.PopulationSize10)
    // assert(
    //   provider.fetch[] .fetch(spark, new Code().withCode("Patient"), None).get
    //     .distinct.count() == 10
    // )
  }

  it should "return exactly 10 bundles when using PopulationSize10" in {
  }

  it should "return None for non-FHIR data types" in {
    assert(sparkcql.retrieve(DataType("http://example.com", "NotFhir")).isEmpty)
  }

  // it should "retrieve encounters when using PopulationSize10" in {
  //   val provider = new FhirSyntheticClinicalDataProvider(FhirSyntheticClinicalDataProvider.PopulationSize10)
  //   assert(
  //     provider10.retrieve(spark, new Code().withCode("Encounter"), None).get
  //       .count() > 100
  //   )
  // }

  // it should "should lazily load" in {
  //   val someNumberBetweenAllPossibleAndLoadedTypes = 5
  //   assert(spark.sql("SHOW TABLES").count() < someNumberBetweenAllPossibleAndLoadedTypes)
  // }
}

File: sparkcql-synthetic/src/test/java/gov/va/sparkcql/synthetic/SyntheaSourceAdapterTest.java
package gov.va.sparkcql.synthetic;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import javax.xml.namespace.QName;

import org.junit.jupiter.api.Test;

import gov.va.sparkcql.Configuration;

import org.apache.spark.sql.SparkSession;
import java.lang.RuntimeException;

public class SyntheaSourceAdapterTest {

    SparkSession spark;
    SyntheaSourceAdapter adapter;

    public SyntheaSourceAdapterTest() {
        this.spark = SparkSession.builder()
            .master("local[1]")
            .getOrCreate();
    }

    @Test
    public void should_dynamically_retrieve_resources() {
        var adapter = new SyntheaSourceAdapter(spark, PopulationSize.PopulationSize10);
        var condition = adapter.acquireData(new QName("http://hl7.org/fhir", "Condition"));
        assertNotNull(condition.head().getAs("id"));
        var encounter = adapter.acquireData(new QName("http://hl7.org/fhir", "Encounter"));
        assertNotNull(encounter.head().getAs("status"));
    }

    @Test
    public void should_load_none_by_default() {
        var builder = new SyntheaDataAdapterFactory();
        var adapter = builder.create(new Configuration(), spark);
        var condition = adapter.acquireData(new QName("http://hl7.org/fhir", "Condition"));
        assertNull(condition);
    }    

    @Test
    public void should_return_none_for_none_fhir_types() {
        assertThrows(
            RuntimeException.class,
            () -> {
                var adapter = new SyntheaSourceAdapter(spark, PopulationSize.PopulationSize10);
                adapter.acquireData(new QName("http://example.com", "NotFhir"));
        });
    }
}